### PR TITLE
GH-96421: Insert shim frame on entry to interpreter

### PR DESCRIPTION
**Pull request by markshannon**
Simplifies `RETURN_VALUE`, `RETURN_GENERATOR` and `YIELD_VALUE` as they no longer need to check if the current frame is the entry frame.

Should allow specialization of `FOR_ITER` and `SEND` for generators and coroutines.

Needs docs and news.

Performance impact is [about zero](https://gist.github.com/markshannon/e618131af12690f3d9eb02a3a8a5933b)

* Issue: gh-96421

### CODE REVIEW

**Comment by sweeneyde** [comment]
Does this add much extra CPU/stack/memory overhead for calling Python from C? Or alternating py/c/py/c/py calls? Probably not the most important use-cases, but I would hope that they wouldn't pessimize *too* much.

File: Lib/test/test_subprocess.py [hunk 1]
```diff
@@ -1875,6 +1876,8 @@ def bad_error(*args):
 
     @unittest.skipIf(not os.path.exists('/proc/self/status'),
                      "need /proc/self/status")
+    @skip_if_sanitizer(memory=True, address=True,
+                       reason= "Spurious error when assigning to stack variable.")
```
Comment (gvanrossum) [inline Lib/test/test_subprocess.py]: Hmm...

File: Objects/frameobject.c [hunk 2]
```diff
@@ -1326,14 +1326,18 @@ PyFrame_LocalsToFast(PyFrameObject *f, int clear)
     }
 }
 
-
 int _PyFrame_IsEntryFrame(PyFrameObject *frame)
```
Comment (gvanrossum) [inline Objects/frameobject.c]:
```suggestion
int
_PyFrame_IsEntryFrame(PyFrameObject *frame)
```

File: Objects/frameobject.c [hunk 3]
```diff
@@ -1326,14 +1326,18 @@ PyFrame_LocalsToFast(PyFrameObject *f, int clear)
...
 {
     assert(frame != NULL);
-    return frame->f_frame->is_entry;
+    _PyInterpreterFrame *f = frame->f_frame;
+    assert(f->owner == FRAME_OWNED_BY_FRAME_OBJECT ||
+           f->owner == FRAME_OWNED_BY_GENERATOR);
+    if (f->previous == NULL) {
+        return 0;
+    }
+    return f->previous->owner == FRAME_OWNED_BY_CSTACK;
```
Comment (gvanrossum) [inline Objects/frameobject.c]:
```suggestion
    return f->previous != NULL && f->previous->owner == FRAME_OWNED_BY_CSTACK;
```

File: Objects/genobject.c [hunk 4]
```diff
@@ -227,11 +225,10 @@ gen_send_ex2(PyGenObject *gen, PyObject *arg, PyObject **presult,
     tstate->exc_info = gen->gi_exc_state.previous_item;
     gen->gi_exc_state.previous_item = NULL;
 
-    assert(tstate->cframe->current_frame == frame->previous);
     /* Don't keep the reference to previous any longer than necessary.  It
      * may keep a chain of frames alive or it could create a reference
      * cycle. */
```
Comment (gvanrossum) [inline Objects/genobject.c]: Update comment.

**Comment by markshannon** [comment]
> Does this add much extra CPU/stack/memory overhead for calling Python from C?

* CPU: A little, but it speeds up the return sequence. The net effect is [about zero](https://gist.github.com/markshannon/e618131af12690f3d9eb02a3a8a5933b)
* C Stack consumption is increased, but only 80 bytes per call.

File: Python/pylifecycle.c [hunk 5]
```diff
@@ -751,6 +752,11 @@ pycore_init_types(PyInterpreterState *interp)
     return _PyStatus_OK();
 }
 
+static const char INTERPRETER_TRAMPOLINE_CODE[] = {
+    0, 0,
+    INTERPRETER_EXIT, 0,
+    RESUME, 0
+};
```
Comment (gvanrossum) [inline Python/pylifecycle.c]: Why is the code separate from the MakeTrampoline function, whereas other details are hardcoded there? Why not hardcode this there as well?

File: Python/pylifecycle.c [hunk 6]
```diff
@@ -751,6 +752,11 @@ pycore_init_types(PyInterpreterState *interp)
     return _PyStatus_OK();
 }
 
+static const char INTERPRETER_TRAMPOLINE_CODE[] = {
+    0, 0,
+    INTERPRETER_EXIT, 0,
```
Comment (sweeneyde) [inline Python/pylifecycle.c:787]: Does this need to worry about endianness?

Comment (gvanrossum) [reply Python/pylifecycle.c:787]: I don't think so, because bytecode is byte-order-independent. On big-endian machines, the opcode is in the high byte; on little-endian machines, it's in the low byte. Either way it's the first byte. :-)

Comment (brandtbucher) [reply Python/pylifecycle.c:787]: Nope! The bytecode format is always opcode, then oparg, regardless of platform.

File: Misc/NEWS.d/next/Core and Builtins/2022-09-09-10-37-34.gh-issue-96421.cyg33z.rst [hunk 7]
```diff
@@ -0,0 +1,4 @@
+When calling into Python code from C code, though PyEval_EvalFrame or
+related C-API function, a shim frame in inserted into the call stack. This
+should be invisible to all Python and most C extensions, but out-of-process
+debuggers, profilers and debuggers need to be aware of this.
```
Comment (brandtbucher) [inline Misc/NEWS.d/next/Core and Builtins/2022-09-09-10-37-34.gh-issue-96421.cyg33z.rst]: This looks like a duplicate of the other NEWS entry. Did you mean to include both?

File: Lib/opcode.py [hunk 8]
```diff
@@ -77,6 +77,7 @@ def pseudo_op(name, op, real_ops):
 def_op('CACHE', 0)
 def_op('POP_TOP', 1)
 def_op('PUSH_NULL', 2)
+def_op('INTERPRETER_EXIT', 3)
```
Comment (brandtbucher) [inline Lib/opcode.py:80]: Maybe just `EXIT`? The `INTERPRETER` part seems implied.

File: Misc/NEWS.d/next/Core and Builtins/2022-10-19-15-59-08.gh-issue-96421.e22y3r.rst [hunk 9]
```diff
@@ -0,0 +1,4 @@
+On entry to ``PyEval_EvalFrameDefault()`` an additional, internal, shim
```
Comment (brandtbucher) [inline Misc/NEWS.d/next/Core and Builtins/2022-10-19-15-59-08.gh-issue-96421.e22y3r.rst]:
```suggestion
On entry to ``_PyEval_EvalFrameDefault()`` an additional, internal, shim
```

File: Misc/NEWS.d/next/Core and Builtins/2022-10-19-15-59-08.gh-issue-96421.e22y3r.rst [hunk 10]
```diff
@@ -0,0 +1,4 @@
...
+frame is pushed. This simplifies the return and yield bytecode instructions.
+Only debuggers and profiles that inspect the stack of frames should be
+impacted.
```
Comment (brandtbucher) [inline Misc/NEWS.d/next/Core and Builtins/2022-10-19-15-59-08.gh-issue-96421.e22y3r.rst]: I assume this only refers to *C* debuggers and profilers that inspect the stack of *Python* frames, (more specifically, the `_PyInterpreterFrame` stack, not the stack of `PyFrameObject`s) right? Probably worth making this explicit to avoid freaking too many people out.

Is it worth mentioning how to detect/skip these frames, too?

File: Objects/frameobject.c [hunk 11]
```diff
@@ -1337,15 +1337,19 @@ PyFrame_LocalsToFast(PyFrameObject *f, int clear)
...
-
 int _PyFrame_IsEntryFrame(PyFrameObject *frame)
 {
     assert(frame != NULL);
-    assert(!_PyFrame_IsIncomplete(frame->f_frame));
-    return frame->f_frame->is_entry;
+    _PyInterpreterFrame *f = frame->f_frame;
+    assert(!_PyFrame_IsIncomplete(f));
+    assert(f->owner == FRAME_OWNED_BY_FRAME_OBJECT ||
+           f->owner == FRAME_OWNED_BY_GENERATOR);
```
Comment (brandtbucher) [inline Objects/frameobject.c]: This used to work on frames owned by the thread, right? Why not anymore?

File: Objects/genobject.c [hunk 12]
```diff
@@ -273,7 +270,6 @@ gen_send_ex2(PyGenObject *gen, PyObject *arg, PyObject **presult,
     _PyErr_ClearExcState(&gen->gi_exc_state);
 
     gen->gi_frame_state = FRAME_CLEARED;
-    _PyFrame_Clear(frame);
```
Comment (brandtbucher) [inline Objects/genobject.c:259]: Maybe add a comment noting (or better, an assertion) that the frame was cleared by the evaluating function.

Also, this slightly changes the contract for PEP 523 extensions, right? All entry frames now are expected to be cleared upon completion, and the `previous` member of generators needs to be cleared on yield? It also affects users of `PyEval_EvalFrame` and `PyEval_EvalFrameEx`.

If so, that seems like it needs a very loud announcement.

File: Python/pylifecycle.c [hunk 13]
```diff
@@ -751,6 +752,11 @@ pycore_init_types(PyInterpreterState *interp)
     return _PyStatus_OK();
 }
 
+static const char INTERPRETER_TRAMPOLINE_CODE[] = {
+    0, 0,
```
Comment (brandtbucher) [inline Python/pylifecycle.c]: Why this `0, 0` pair?

File: Python/pylifecycle.c [hunk 14]
```diff
@@ -751,6 +752,11 @@ pycore_init_types(PyInterpreterState *interp)
     return _PyStatus_OK();
 }
 
+static const char INTERPRETER_TRAMPOLINE_CODE[] = {
```
Comment (brandtbucher) [inline Python/pylifecycle.c]:
```suggestion
static const uint8_t INTERPRETER_TRAMPOLINE_CODE[] = {
```

I think we should probably be better about using explicit, unsigned values for bytecode, since I'm pretty sure that the signed overflow from assigning an opcode/oparg with value >= 128 is technically undefined.

Also, consider lowercase for `INTERPRETER_TRAMPOLINE_CODE`? I get that it's a constant, but our style typically reserves uppers for macros.

File: Tools/gdb/libpython.py [hunk 15]
```diff
@@ -1078,7 +1078,7 @@ def _f_lasti(self):
         return int(prev_instr - first_instr)
 
     def is_entry(self):
-        return self._f_special("is_entry", bool)
+        return self._f_special("owner", int) == 3
```
Comment (brandtbucher) [inline Tools/gdb/libpython.py]: Maybe name `3` as `FRAME_OWNED_BY_CSTACK`, so find-and-replace, etc. picks this up?

Also, this is a tiny bit confusing, because we have a different definition of an "entry frame" in `_PyFrame_IsEntryFrame`. Maybe use that definition, and leave the `break` lines where they are? Or update this to be `is_shim` or `is_trampoline`.

File: Python/ceval.c [hunk 16]
```diff
@@ -155,7 +155,10 @@ static void
 lltrace_resume_frame(_PyInterpreterFrame *frame)
 {
     PyObject *fobj = frame->f_funcobj;
-    if (fobj == NULL || !PyFunction_Check(fobj)) {
+    if (frame->owner == FRAME_OWNED_BY_CSTACK ||
+        fobj == NULL ||
+        !PyFunction_Check(fobj)
```
Comment (brandtbucher) [inline Python/ceval.c:160]: For my own understanding, when is `fobj` `NULL` (or a non-function)? I don't see how this can happen, and there are even a few places where we use `f_funcobj` without asserting or checking for `NULL` first.

File: Python/ceval.c [hunk 17]
```diff
@@ -1057,9 +1053,23 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
...
 
-    frame->is_entry = true;
+    assert(tstate->interp->interpreter_trampoline != NULL);
+#ifdef Py_DEBUG
+    pyframe.f_funcobj = (PyObject*)0xaaa0;
+    pyframe.f_locals = (PyObject*)0xaaa1;
+    pyframe.frame_obj = (PyFrameObject*)0xaaa2;
+    pyframe.f_globals = (PyObject*)0xaaa3;
+    pyframe.f_builtins = (PyObject*)0xaaa4;
+#endif
```
Comment (brandtbucher) [inline Python/ceval.c]: Maybe add a comment that this is (presumably) for debugging the inevitable segfaulting dereferences in the future? ;)

File: Python/ceval.c [hunk 18]
```diff
@@ -1057,9 +1053,23 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
...
+    pyframe.f_code = tstate->interp->interpreter_trampoline;
+    _Py_CODEUNIT *code = _PyCode_CODE(tstate->interp->interpreter_trampoline);
+    pyframe.prev_instr = code;
```
Comment (brandtbucher) [inline Python/ceval.c]:
```suggestion
    pyframe.prev_instr = _PyCode_CODE(tstate->interp->interpreter_trampoline);
```

File: Python/ceval.c [hunk 19]
```diff
@@ -1045,8 +1041,8 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
 #endif
 
     _PyCFrame cframe;
-    CallShape call_shape;
-    call_shape.kwnames = NULL; // Borrowed reference. Reset by CALL instructions.
+    _PyInterpreterFrame  pyframe;
```
Comment (brandtbucher) [inline Python/ceval.c]:
```suggestion
    _PyInterpreterFrame pyframe;
```

Also, consider renaming to `shim_frame` or `trampoline_frame`. A name like `pyframe` is way too overloaded to actually be helpful.

File: Include/internal/pycore_frame.h [hunk 20]
```diff
@@ -42,27 +42,28 @@ typedef enum _framestate {
...
     // NOTE: This is not necessarily the last instruction started in the given
     // frame. Rather, it is the code unit *prior to* the *next* instruction. For
     // example, it may be an inline CACHE entry, an instruction we just jumped
     // over, or (in the case of a newly-created frame) a totally invalid value:
     _Py_CODEUNIT *prev_instr;
-    int stacktop;     /* Offset of TOS from localsplus  */
-    bool is_entry;  // Whether this is the "root" frame for the current _PyCFrame.
-    char owner;
+    int stacktop;  /* Offset of TOS from localsplus  */
+    int yield_offset: 28;
```
Comment (brandtbucher) [inline Include/internal/pycore_frame.h]: Is this actually being used yet? Might make more sense to wait for a later PR, if not.

File: Include/internal/pycore_frame.h [hunk 21]
```diff
@@ -42,27 +42,28 @@ typedef enum _framestate {
...
-    PyObject *f_globals; /* Borrowed reference */
-    PyObject *f_builtins; /* Borrowed reference */
-    PyObject *f_locals; /* Strong reference, may be NULL */
+    PyObject *f_funcobj; /* Strong reference. Only valid for Python frames */
+    PyObject *f_globals; /* Borrowed reference. Only valid for Python frames */
+    PyObject *f_builtins; /* Borrowed reference. Only valid for Python frames */
+    PyObject *f_locals; /* Strong reference, may be NULL. Only valid for Python frames */
     PyCodeObject *f_code; /* Strong reference */
-    PyFrameObject *frame_obj; /* Strong reference, may be NULL */
+    PyFrameObject *frame_obj; /* Strong reference, may be NULL. Only valid for Python frames */
```
Comment (brandtbucher) [inline Include/internal/pycore_frame.h]: Maybe clarify "Python frame" here? If I understand correctly, these members are "Only valid for complete frames that are not owned by the C stack"?

File: Python/ceval.c [hunk 22]
```diff
@@ -1118,14 +1126,16 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
 
 #ifdef LLTRACE
     {
-        int r = PyDict_Contains(GLOBALS(), &_Py_ID(__lltrace__));
-        if (r < 0) {
-            goto exit_unwind;
+        if (frame != &pyframe) {
```
Comment (brandtbucher) [inline Python/ceval.c]: It probably still makes sense to trace shim frames, right? If you have `__lltrace__` defined, you *really* do care about things like `INTERPRETER_EXIT`.

File: Objects/genobject.c [hunk 23]
```diff
@@ -273,7 +270,6 @@ gen_send_ex2(PyGenObject *gen, PyObject *arg, PyObject **presult,
     _PyErr_ClearExcState(&gen->gi_exc_state);
 
     gen->gi_frame_state = FRAME_CLEARED;
```
Comment (brandtbucher) [inline Objects/genobject.c]: This already happened, right? Maybe turn this into an assertion too?

File: Python/ceval.c [hunk 24]
```diff
@@ -2069,20 +2085,19 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
...
-            /* Restore previous cframe and return. */
-            tstate->cframe = cframe.previous;
-            tstate->cframe->use_tracing = cframe.use_tracing;
-            assert(tstate->cframe->current_frame == frame->previous);
-            assert(!_PyErr_Occurred(tstate));
-            return retval;
+            _PyInterpreterFrame *gen_frame = frame;
+            frame = cframe.current_frame = gen_frame->previous;
+            gen_frame->previous = NULL;
+            frame->prev_instr += frame->yield_offset;
```
Comment (brandtbucher) [inline Python/ceval.c]: Just checking, this is always zero, right?

File: Python/frame.c [hunk 25]
```diff
@@ -139,10 +142,12 @@ _PyFrame_Clear(_PyInterpreterFrame *frame)
     for (int i = 0; i < frame->stacktop; i++) {
         Py_XDECREF(frame->localsplus[i]);
     }
+    frame->stacktop = 0;
```
Comment (brandtbucher) [inline Python/frame.c]: Wait, should this line be backported? It seems like it could lead to trouble if omitted.

File: Python/pylifecycle.c [hunk 26]
```diff
@@ -751,6 +752,11 @@ pycore_init_types(PyInterpreterState *interp)
...
+    0, 0,
+    INTERPRETER_EXIT, 0,
+    RESUME, 0
```
Comment (brandtbucher) [inline Python/pylifecycle.c:789]: Why is this here?

File: Python/traceback.c [hunk 27]
```diff
@@ -1229,6 +1229,13 @@ dump_traceback(int fd, PyThreadState *tstate, int write_header)
         if (frame == NULL) {
             break;
         }
+        if (frame->owner == FRAME_OWNED_BY_CSTACK) {
+            /* Trampoline frame */
+            frame = frame->previous;
+        }
```
Comment (brandtbucher) [inline Python/traceback.c:1230]: It's never possible to have two trampolines in a row, right? I'm thinking of some nightmare scenario where new code starts running during a trampoline frame for some reason (like as part of a `Py_DECREF`, which I don't think can happen).

File: Objects/codeobject.c [hunk 28]
```diff
@@ -2268,3 +2268,57 @@ _PyStaticCode_InternStrings(PyCodeObject *co)
     }
     return 0;
 }
+
+PyCodeObject *
+_Py_MakeTrampoline(const char *code, int codelen, int stacksize, const char *cname)
```
Comment (brandtbucher) [inline Objects/codeobject.c]: So, this interface feels a bit strange to me. We only ever create one trampoline code object, right? Maybe take the `INTERPRETER_TRAMPOLINE_CODE` and all of the other arguments out of `pylifecycle.c` and move them here, since there's a pretty tight coupling between the arguments and the rest of the new code members.

Also, `codelen` should probably be the number of code units, not the number of bytes.

File: Objects/codeobject.c [hunk 29]
```diff
@@ -2268,3 +2268,57 @@ _PyStaticCode_InternStrings(PyCodeObject *co)
...
+
+    name = _PyUnicode_FromASCII(cname, strlen(cname));
+    if (name == NULL) {
+        goto cleanup;
+    }
+    co_code = PyBytes_FromStringAndSize(code, codelen);
+    if (co_code == NULL) {
+        goto cleanup;
+    }
+    const char loc[2] = { 0x80 | (PY_CODE_LOCATION_INFO_NO_COLUMNS << 3) | 3, 0 };
```
Comment (brandtbucher) [inline Objects/codeobject.c]: For example, this linetable makes some pretty serious assumptions about the size `code`/`codelen`.

Also, I think it's wrong in this case (`INTERPRETER_TRAMPOLINE_CODE` has three code units, but this table describes locations for four code units).

File: Python/pylifecycle.c [hunk 30]
```diff
@@ -785,7 +791,11 @@ pycore_init_builtins(PyThreadState *tstate)
     PyObject *object__getattribute__ = _PyType_Lookup(&PyBaseObject_Type, &_Py_ID(__getattribute__));
     assert(object__getattribute__);
     interp->callable_cache.object__getattribute__ = object__getattribute__;
-
+    interp->interpreter_trampoline = _Py_MakeTrampoline(
+        INTERPRETER_TRAMPOLINE_CODE, sizeof(INTERPRETER_TRAMPOLINE_CODE), 1, "<interpreter trampoline>");
```
Comment (brandtbucher) [inline Python/pylifecycle.c]: See my other comment. `_Py_MakeTrampoline` is pretty tightly coupled to these exact argument values, so they probably just belong there.

**Review by brandtbucher — CHANGES_REQUESTED** [review]
Thanks for taking the time to do this. There's definitely a lot of subtlety involved.

In general, I think that this adds a fair amount of complexity that wasn't present before (I've recently gained new appreciation for the value of being able to reason about the web of frames that exist in 3.11+, particularly things like visibility and ownership rules). If you're confident that we can recoup the perf cost and complexity with work that builds upon this, though, I think it makes sense as a first step towards that.

Also, I think that the growing set of rules around things like "ownership", "completeness", and "state" are becoming a real maintenance burden. We might consider tightening up those concepts in 3.12, because there are lots of complex relationships and invalid states that are crucial to understand when working on (or with) interpreter internals. I'm thinking not only about us, but also third-party libraries like Greenlet and forks like Cinder.

File: Objects/codeobject.c (hunk 28 above)
Comment (brandtbucher) [reply Objects/codeobject.c]: Ah, maybe it's meant for reuse with future cleanup frames? In that case, it should be fine to just take a single opcode as an argument, right (`INTERPRETER_EXIT` in this case, maybe `CLEANUP_INIT` or `CLEANUP_SEND` in the future)?

File: Objects/codeobject.c (hunk 29 above)
Comment (brandtbucher) [reply Objects/codeobject.c]: (Also, another place we should probably use `uint8_t`.)

File: Python/ceval.c (hunk 22 above)
Comment (brandtbucher) [reply Python/ceval.c]: Ah, wait, but exceptions can’t be raised in shim frames. Hm.

File: Include/internal/pycore_frame.h (hunk 21 above)
Comment (markshannon) [reply Include/internal/pycore_frame.h]: C extensions could insert frames, for profilers and such. Although, this is hypothetical at this point.

Those fields are valid for incomplete frames. I'll go with "Only valid if not on C stack" for now.

File: Lib/opcode.py (hunk 8 above)
Comment (markshannon) [reply Lib/opcode.py:80]: IMO `EXIT` sounds like it exits the program, like `HALT`.

File: Misc/NEWS.d/next/Core and Builtins/2022-09-09-10-37-34.gh-issue-96421.cyg33z.rst (hunk 7 above)
Comment (markshannon) [reply Misc/NEWS.d/next/Core and Builtins/2022-09-09-10-37-34.gh-issue-96421.cyg33z.rst]: No, one should be enough :slightly_smiling_face:

File: Objects/codeobject.c (hunk 28 above)
Comment (markshannon) [reply Objects/codeobject.c]: I expect to generalize it for specializing calls to classes, so the shim function can potentially do more work than that.
It makes sense to bundle the code, length, etc into some sort of struct.

File: Python/frame.c (hunk 25 above)
Comment (markshannon) [reply Python/frame.c]: I probably added this when debugging. `_PyFrame_Clear` is called when the frame is being destroyed, so it doesn't matter. I'll remove this and the `frame->owner = FRAME_CLEARED;` below.

File: Python/pylifecycle.c (hunk 14 above)
Comment (markshannon) [reply Python/pylifecycle.c]: We use ALL_CAPS for constants as well, although we aren't consistent about it.

File: Python/pylifecycle.c (hunk 13 above)
Comment (markshannon) [reply Python/pylifecycle.c]: I'ved changed the first `0` to a `NOP`. The idea is that the instruction pointer points *into* the code, as if it were already executing.

File: Python/pylifecycle.c (hunk 26 above)
Comment (markshannon) [reply Python/pylifecycle.c:789]: So that this frame is incomplete. I'll add comments.

File: Python/traceback.c (hunk 27 above)
Comment (markshannon) [reply Python/traceback.c:1230]: I don't see how this can happen, but I'll add an assert.

File: Python/ceval.c (hunk 24 above)
Comment (markshannon) [reply Python/ceval.c]: Yes, but I've removed it for now

File: Objects/frameobject.c (hunk 11 above)
Comment (markshannon) [reply Objects/frameobject.c]: Probably a left over from debugging.

File: Python/ceval.c (hunk 16 above)
Comment (markshannon) [reply Python/ceval.c:160]: This is for debugging the interpreter. So even though `fobj` shouldn't be `NULL`, we want this to mostly work even if `fobj == NULL`.

File: Python/ceval.c (hunk 22 above)
Comment (markshannon) [reply Python/ceval.c]: You can't lookup `__lltrace__` in the globals, if you don't have any globals.

File: Objects/genobject.c (hunk 12 above)
Comment (markshannon) [reply Objects/genobject.c:259]: Actually it makes little difference for users of PEP 523.
There are two types of users of PEP 523.
* Those that tweak the frame in some way and call `_PyEval_EvalDefault()`. It makes no difference to them as they did no cleanup anyway.
* Those that reimplement the interpreter (I suspect this type is mythical). They need to fully re-implement everything for every release anyway.

`PyEval_EvalFrame` and `PyEval_EvalFrameEx` are not changed because they take frame objects. So when the frame is to be cleared, the frame object takes ownership of it.

**Comment by markshannon** [comment]
I have made the requested changes; please review again

**Comment by brandtbucher** [comment]
I don't know if I'll have time to review again today (still need to prepare my talk for the release stream). Is this able to wait until next week? If not, I can try to make time.

**Comment by markshannon** [comment]
It can wait until next week

File: Misc/NEWS.d/next/Core and Builtins/2022-10-19-15-59-08.gh-issue-96421.e22y3r.rst [hunk 31]
```diff
@@ -0,0 +1,14 @@
+When calling into Python code from C code, though ``PyEval_EvalFrame`` or
+related C-API function, a shim frame in inserted into the call stack.
+This occurs in the ``_PyEval_EvalFrameDefault()`` function.
+The extra frame should be invisible to all Python and most C extensions,
+but out-of-process debuggers, profilers and debuggers need to be aware of
```
Comment (brandtbucher) [inline Misc/NEWS.d/next/Core and Builtins/2022-10-19-15-59-08.gh-issue-96421.e22y3r.rst]:
```suggestion
but out-of-process profilers and debuggers need to be aware of
```

File: Objects/codeobject.c [hunk 32]
```diff
@@ -2268,3 +2268,76 @@ _PyStaticCode_InternStrings(PyCodeObject *co)
...
+    name = _PyUnicode_FromASCII(codedef->cname, strlen(codedef->cname));
+    if (name == NULL) {
+        goto cleanup;
+    }
+    co_code = PyBytes_FromStringAndSize(
+        (const char *)codedef->code, codedef->codelen);
+    if (co_code == NULL) {
+        goto cleanup;
+    }
+    int code_units = codedef->codelen/2;
```
Comment (brandtbucher) [inline Objects/codeobject.c]:
```suggestion
    int code_units = codedef->codelen / sizeof(_Py_CODEUNIT)
```

File: Objects/codeobject.c [hunk 33]
```diff
@@ -2268,3 +2268,76 @@ _PyStaticCode_InternStrings(PyCodeObject *co)
...
+    int loc_entries = (code_units + 7)/8;
```
Comment (brandtbucher) [inline Objects/codeobject.c]: Maybe add a comment like "The maximum number of code units encoded by a single line table entry is 8. This is equivalent to `ceil(code_units / 8)`."

File: Objects/codeobject.c [hunk 34]
```diff
@@ -2268,3 +2268,76 @@ _PyStaticCode_InternStrings(PyCodeObject *co)
...
+    loc_table = PyMem_Malloc(loc_entries);
+    if (loc_table == NULL) {
+        PyErr_NoMemory();
+        return NULL;
```
Comment (brandtbucher) [inline Objects/codeobject.c]:
```suggestion
        goto cleanup;
```

File: Objects/codeobject.c [hunk 35]
```diff
@@ -2268,3 +2268,76 @@ _PyStaticCode_InternStrings(PyCodeObject *co)
...
+        .stacksize = codedef->stacksize,
+
+        .exceptiontable = (PyObject *)&_Py_SINGLETON(bytes_empty),
+    };
+
+    codeobj = _PyCode_New(&con);
+cleanup:
+    if (loc_table) {
+        PyMem_Free(loc_table);
+    }
```
Comment (brandtbucher) [inline Objects/codeobject.c]: This can be done above without the conditional.

```suggestion
```

File: Objects/codeobject.c [hunk 36]
```diff
@@ -2268,3 +2268,76 @@ _PyStaticCode_InternStrings(PyCodeObject *co)
...
+    for (int i = 0; i < loc_entries-1; i++) {
+         loc_table[i] = 0x80 | (PY_CODE_LOCATION_INFO_NONE << 3) | 7;
+         code_units -= 8;
+    }
+    assert(loc_entries > 0);
+    assert(code_units > 0 && code_units <= 8);
+    loc_table[loc_entries-1] = 0x80 |
+        (PY_CODE_LOCATION_INFO_NONE << 3) | (code_units-1);
+    lines = PyBytes_FromStringAndSize((const char *)loc_table, loc_entries);
```
Comment (brandtbucher) [inline Objects/codeobject.c:2304]:
```suggestion
    lines = PyBytes_FromStringAndSize((const char *)loc_table, loc_entries);
    PyMem_Free(loc_table);
```

File: Objects/codeobject.c [hunk 37]
```diff
@@ -2268,3 +2268,76 @@ _PyStaticCode_InternStrings(PyCodeObject *co)
...
+    if (lines == NULL) {
+        goto cleanup;
+    }
+    struct _PyCodeConstructor con = {
+        .filename = &_Py_STR(empty),
```
Comment (brandtbucher) [inline Objects/codeobject.c]: Not a *huge* deal, but I think we typically put *something* here for ease of debugging. Maybe `"<internal>"`?

**Review by brandtbucher — COMMENTED** [review]
Mostly nits on the changes.

Also, I'd like to discuss this vs. something like #98795 again at our stand-up, if that's okay. This still feels pretty heavy/breaky to me compared to that PR (and I'm still not clear what it actually gives us over that approach).

File: Python/pylifecycle.c (hunk 26 above)
Comment (brandtbucher) [reply Python/pylifecycle.c:789]: I don't think it's needed, though. `co->_co_firsttraceable == Py_SIZE(co)` if no `RESUME` is present, right?

Comment (markshannon) [reply Python/pylifecycle.c:789]: Technically yes, but that's really an implementation detail. The rule is that a frame is invalid until `RESUME` or `RETURN_GENERATOR` has been reached.

File: Python/ceval.c [hunk 38]
```diff
@@ -1852,27 +1862,33 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
             goto error;
         }
 
-        TARGET(RETURN_VALUE) {
```
Comment (gvanrossum) [inline Python/ceval.c]: Reminder that you'll get a merge conflict here -- instruction changes now need to be in bytecodes.c and you need to run `make regen-cases`.

**Review by gvanrossum — COMMENTED** [review]
I had some earlier comments pending (not sure if they still apply). But mainly this is a reminder that instructions are now in bytecodes.c, and changes there require running `make regen-cases`.

File: Python/pylifecycle.c (hunk 5 above)
Comment (markshannon) [reply Python/pylifecycle.c]: The `MakeTrampoline`function is for making trampolines in general. This defines the specific trampoline for exiting `PyEval_EvalDefault()`

File: Lib/test/test_subprocess.py [hunk 39]
```diff
@@ -25,6 +25,7 @@
 import json
 import pathlib
 from test.support.os_helper import FakePath
+from test.support import skip_if_sanitizer
```
Comment (brandtbucher) [inline Lib/test/test_subprocess.py]:
```suggestion
```

File: Misc/NEWS.d/next/Core and Builtins/2022-10-19-15-59-08.gh-issue-96421.e22y3r.rst [hunk 40]
```diff
@@ -0,0 +1,13 @@
...
+This occurs in the ``_PyEval_EvalFrameDefault()`` function.
+The extra frame should be invisible to all Python and most C extensions,
+but out-of-process profilers and debuggers need to be aware of it.
+These shim frames can be detected by checking
+``frame->owner == FRAME_OWNED_BY_CSTACK``.
+
+Extensions implementing their own interpreters using PEP 523 need to be
+aware of this shim frame and the changes to the semantics of
+``RETURN_VALUE``, ``YIELD_VALUE``, and ``RETURN_GENERATOR``,
+which now clear the frame.
```
Comment (brandtbucher) [inline Misc/NEWS.d/next/Core and Builtins/2022-10-19-15-59-08.gh-issue-96421.e22y3r.rst]:
```suggestion
When calling into Python code from C code, through :c:func:`PyEval_EvalFrameEx` or
a related C-API function, a shim frame in inserted into the call stack.
This occurs in the ``_PyEval_EvalFrameDefault()`` function.
The extra frame should be invisible to all Python and most C extensions,
but out-of-process profilers and debuggers need to be aware of it.
These shim frames can be detected by checking
``frame->owner == FRAME_OWNED_BY_CSTACK``.

Extensions implementing their own interpreters using PEP 523 need to be
aware of this shim frame and the changes to the semantics of
:opcode:`RETURN_VALUE`, :opcode:`YIELD_VALUE`, and :opcode:`RETURN_GENERATOR`,
which now clear the frame.
```

File: Lib/opcode.py (hunk 8 above)
Comment (brandtbucher) [inline Lib/opcode.py:80]: `INTERPRETER_EXIT` needs to be documented in `dis.rst`.

**Review by brandtbucher — APPROVED** [review]
Just a couple of remaining nits.

I'm not sure what kind of review you're looking for... I don't see any technical issues with the implementation, so I'll hit "approve". :)

But I still have reservations about this change being much deeper (and probably more expensive) than what's really necessary right now. I still feel like I haven't *really* gotten an answer on why this change is more appropriate than something like #98795. So I don't necessarily *approve* of this approach, but I also won't block you from merging this if you still feel differently.

File: Lib/opcode.py (hunk 8 above)
Comment (markshannon) [reply Lib/opcode.py:80]: `INTERPRETER_EXIT` can never appear in any code object (apart from the shim, and that is inaccessible) so cannot appear in dis output.
